atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/store/AtlasDbTransactionStore.java

```diff
@@ -99,9 +113,22 @@ public Optional<WitnessedTransaction> readWrite(Consumer<InteractiveTransaction>
                     .build());
         } catch (SafeIllegalArgumentException e) {
             throw e;
+        } catch (KeyAlreadyExistsException e) {
+            Transaction transaction = transactionReference.get();
+            return Optional.of(MaybeWitnessedTransaction.builder()
+                    .startTimestamp(transaction.getTimestamp())
+                    .commitTimestamp(commitTimestampProvider
+                            .get()
+                            .getCommitTimestampOrThrowIfMaybeNotCommitted(transaction.getTimestamp()))
+                    .actions(witnessedActionsReference.get())
+                    .build());
         } catch (Exception e) {
-            // TODO: Need to eventually handle PuE exceptions, as they could've succeeded in committing.
-            log.info("Failed to record transaction due to an exception", e);
+            Optional<Long> startTimestamp =
+                    Optional.ofNullable(transactionReference.get()).map(Transaction::getTimestamp);
+            log.info(
+                    "Failed to record transaction due to an exception for startTimestamp {}",
+                    SafeArg.of("startTimestamp", startTimestamp),
+                    e);
```

Comment:
We might want to set the transaction reference at the beginning of the task, if we're concerned about something going wrong before we enter the commit-writes phase.